browser/ui/webui/settings/brave_default_extensions_handler.cc

```diff
@@ -44,6 +55,49 @@ void BraveDefaultExtensionsHandler::RegisterMessages() {
       base::BindRepeating(
         &BraveDefaultExtensionsHandler::SetIPFSCompanionEnabled,
         base::Unretained(this)));
+  web_ui()->RegisterMessageCallback(
+      "setMediaRouterEnabled",
+      base::BindRepeating(
+        &BraveDefaultExtensionsHandler::SetMediaRouterEnabled,
+        base::Unretained(this)));
+  web_ui()->RegisterMessageCallback(
+      flags_ui::kRestartBrowser,
+      base::BindRepeating(&BraveDefaultExtensionsHandler::HandleRestartBrowser,
+                          base::Unretained(this)));
+  web_ui()->RegisterMessageCallback(
+      "getRestartNeeded",
+      base::BindRepeating(
+        &BraveDefaultExtensionsHandler::GetRestartNeeded,
+        base::Unretained(this)));
+}
+
+void BraveDefaultExtensionsHandler::GetRestartNeeded(
+    const base::ListValue* args) {
+  bool media_router_current_pref = profile_->GetPrefs()->GetBoolean(
+      prefs::kEnableMediaRouter);
+  bool media_router_new_pref = profile_->GetPrefs()->GetBoolean(
+      kBraveEnabledMediaRouter);
+  bool media_router_enabled = extensions::ExtensionPrefs::Get(profile_)->
+      IsExtensionDisabled(extension_misc::kMediaRouterStableExtensionId);
+
+  // This only happens on upgrade if the user has enabled media router using
+  // brave://flags before. We explicitly require the user to enable the pref
+  // to avoid any surprises about new connections on upgrade.
+  if (media_router_enabled != media_router_current_pref) {
+    restart_needed_ = (media_router_current_pref == media_router_new_pref);
+  } else {
+    restart_needed_ = (media_router_current_pref != media_router_new_pref);
+  }
+
+
+  AllowJavascript();
+  ResolveJavascriptCallback(args->GetList()[0].Clone(),
+      base::Value(restart_needed_));
+}
+
+void BraveDefaultExtensionsHandler::HandleRestartBrowser(
```

Comment:
I believe the correct way to handle this is to just link to `kChromeUIRestartURL`

Reply:
Navigation to `kChromeUIRestartURL` internally calls `chrome::AttemptRestart`: https://cs.chromium.org/chromium/src/chrome/browser/browser_about_handler.cc?l=98.

We can link to `kChromeUIRestartURL` in javascript directly but I was just trying to avoid chances of regression if `kChromeUIRestartURL` changes in future.

Reply:
I think it is better to use `kChromeUIRestartURL` than to duplicate the logic here. Not sure why `kChromeUIRestartURL` would change in the future

Reply:
@bridiver just noting that the same functionality in `chrome://flags` does not link to `kChromeUIRestartUI` but [sends a message to a message handler via `chrome.send` too](https://cs.chromium.org/chromium/src/components/flags_ui/resources/flags.js?type=cs&q=experiment-restart-button&sq=package:chromium&g=0&l=139) and [calls `AttemptRestart`](https://cs.chromium.org/chromium/src/chrome/browser/ui/webui/flags_ui.cc?type=cs&sq=package:chromium&g=0&l=268). I wonder what would happen to session restore if the tab was navigated away from the settings page to `kChromeUIRestartURL`. The settings page may not restore, which would be unfortunate. Although I did spot comments around this being a non-navigation url, so maybe it would restore... Still, I wonder why the Flags page doesn't use it.

Reply:
@petemill because a lot of other things happen in `FlagsDOMHandler::HandleRestartBrowser` before the restart. The only thing we're doing here is calling `AttemptRestart` so it's functionally equivalent

Reply:
```void FlagsDOMHandler::HandleRestartBrowser(const base::ListValue* args) {
  DCHECK(flags_storage_);
#if defined(OS_CHROMEOS)
  // On ChromeOS be less intrusive and restart inside the user session after
  // we apply the newly selected flags.
  base::CommandLine user_flags(base::CommandLine::NO_PROGRAM);
  about_flags::ConvertFlagsToSwitches(flags_storage_.get(),
                                      &user_flags,
                                      flags_ui::kAddSentinels);

  // Apply additional switches from policy that should not be dropped when
  // applying flags..
  chromeos::UserSessionManager::MaybeAppendPolicySwitches(
      Profile::FromWebUI(web_ui())->GetPrefs(), &user_flags);

  base::CommandLine::StringVector flags;
  // argv[0] is the program name |base::CommandLine::NO_PROGRAM|.
  flags.assign(user_flags.argv().begin() + 1, user_flags.argv().end());
  VLOG(1) << "Restarting to apply per-session flags...";
  AccountId account_id =
      user_manager::UserManager::Get()->GetActiveUser()->GetAccountId();
  chromeos::UserSessionManager::GetInstance()->SetSwitchesForUser(
      account_id,
      chromeos::UserSessionManager::CommandLineSwitchesType::
          kPolicyAndFlagsAndKioskControl,
      flags);
#endif
  chrome::AttemptRestart();
}
```

Reply:
they need all of that code to run on chromeos before attempting the restart